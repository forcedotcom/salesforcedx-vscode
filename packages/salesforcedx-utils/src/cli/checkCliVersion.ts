/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { execSync } from 'child_process';
import * as semver from 'semver';

export enum CheckCliEnum {
  validCli = 1,
  outdatedSFDXVersion = 2,
  onlySFv1 = 3,
  cliNotInstalled = 4,
  bothSFDXAndSFInstalled = 5
}

export class CheckCliVersion {

  public getSfdxCliVersion(): string {
    try {
      const result = execSync('sfdx --version');
      return result.toString();
    } catch {
      return 'No SFDX CLI';
    }
  }
  public getSfCliVersion(): string {
    try {
      const result = execSync('sf --version');
      return result.toString();
    } catch {
      return 'No SF CLI';
    }
  }

  public parseCliVersion(sfCliVersion: string): string {
    const pattern = (?:sfdx-cli\/|@salesforce\/cli\/)(\d+\.\d+\.\d+);
    const match = pattern.exec(sfCliVersion);
    // SFDX v7 reports results in match[1], SF v2 reports results in match[2]
    return match ? match[1] : '0.0.0';
  }

  public validateCliInstallationAndVersion(sfdxCliVersionString: string, sfCliVersionString: string): CheckCliEnum {

    // Case 1: Neither SFDX CLI nor SF CLI is installed
    if (semver.satisfies(sfdxCliVersionString, '0.0.0') && semver.satisfies(sfCliVersionString, '0.0.0')) {
      return CheckCliEnum.cliNotInstalled;
    }

    // Case 2: Only SF CLI (v1) is installed (SF v1 cannot be used because it does not map sf to sfdx)
    if (semver.satisfies(sfdxCliVersionString, '0.0.0') && semver.satisfies(sfCliVersionString, '1.x')) {
      return CheckCliEnum.onlySFv1;
    }

    // Case 3: Both SFDX CLI (v7) and SF CLI (v2) are installed at the same time
    if (semver.satisfies(sfCliVersionString, '2.x') && !semver.satisfies(sfdxCliVersionString, sfCliVersionString)) {
      return CheckCliEnum.bothSFDXAndSFInstalled;
    }

    // Case 4: Outdated SFDX CLI version is installed
    const minSFDXVersion = '7.193.2';
    if (semver.satisfies(sfdxCliVersionString, (`<${minSFDXVersion}`) && semver.satisfies(sfCliVersionString, '<2.0.0')) {
      return CheckCliEnum.outdatedSFDXVersion;
    }

    // Case 5: Valid SFDX v7 version or SF v2 is installed
    return CheckCliEnum.validCli;
  }
}
