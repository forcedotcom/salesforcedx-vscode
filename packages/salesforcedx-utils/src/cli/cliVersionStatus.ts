/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { execSync } from 'child_process';
import * as semver from 'semver';

export enum CliStatusEnum {
  SFv2 = 1,
  outdatedSFDXVersion = 2,
  onlySFv1 = 3,
  cliNotInstalled = 4,
  bothSFDXAndSFInstalled = 5,
  SFDXv7Valid = 6
}

export class CliVersionStatus {

  public getCliVersion(isSfdx: boolean): string {
    try {
      const result = execSync(`${isSfdx ? 'sfdx' : 'sf'} --version`);
      return result.toString();
    } catch {
      return 'No CLI';
    }
  }

  public parseCliVersion(cliVersion: string): string {
    const pattern = /(?:sfdx-cli\/|@salesforce\/cli\/)(\d+\.\d+\.\d+)/;
    const match = pattern.exec(cliVersion);
    return match ? match[1] : '0.0.0';
  }

  public validateCliInstallationAndVersion(sfdxCliVersionString: string, sfCliVersionString: string): CliStatusEnum {

    // Case 1: Neither SFDX CLI nor SF CLI is installed
    if (semver.satisfies(sfdxCliVersionString, '0.0.0') && semver.satisfies(sfCliVersionString, '0.0.0')) {
      return CliStatusEnum.cliNotInstalled;
    }

    // Case 2: Only SF CLI (v1) is installed (SF v1 cannot be used because it does not map sf to sfdx)
    if (semver.satisfies(sfdxCliVersionString, '0.0.0') && semver.satisfies(sfCliVersionString, '1.x')) {
      return CliStatusEnum.onlySFv1;
    }

    // Case 3: Both SFDX CLI (v7) and SF CLI (v2) are installed at the same time
    if (semver.satisfies(sfCliVersionString, '2.x') && !semver.satisfies(sfdxCliVersionString, sfCliVersionString)) {
      return CliStatusEnum.bothSFDXAndSFInstalled;
    }

    const minSFDXVersion = '7.193.2';
    if (semver.satisfies(sfCliVersionString, '<2.0.0')) {
      if (semver.satisfies(sfdxCliVersionString, (`<${minSFDXVersion}`))) {
        // Case 4: Outdated SFDX CLI version is installed
        return CliStatusEnum.outdatedSFDXVersion;
      } else {
        // Case 5: Valid SFDX v7 version is installed
        return CliStatusEnum.SFDXv7Valid;
      }
    }

    // Case 6: SF v2 is installed
    return CliStatusEnum.SFv2;
  }
}
