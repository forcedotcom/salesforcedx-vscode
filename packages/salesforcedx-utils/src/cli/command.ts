import { CommandBuilder } from './commandBuilder';

export class Command {
  public readonly command: string;
  public readonly description?: string;
  public readonly args: string[];
  public readonly logName?: string;

  public constructor(builder: CommandBuilder) {
    this.command = builder.command;
    this.description = builder.description;
    this.args = builder.args;
    this.logName = builder.logName;
  }

  public toString(): string {
    return this.description
      ? this.description
      : `${this.command} ${this.args.join(' ')}`;
  }

  public toCommand(): string {
    return `${this.command} ${this.args.join(' ')}`;
  }
}
