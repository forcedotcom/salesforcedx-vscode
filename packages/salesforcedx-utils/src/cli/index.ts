export { CliCommandExecution } from './cliCommandExecution';
export { CliCommandExecutor } from './cliCommandExecutor';
export { CommandBuilder } from './commandBuilder';
export { CommandOutput } from './commandOutput';
export { ForceOrgDisplay } from './forceOrgDisplay';
export { GlobalCliEnvironment } from './globalCliEnvironment';
export { SfdxCommandBuilder } from './sfdxCommandBuilder';
export { ForceConfigGet } from './forceConfigGet';
