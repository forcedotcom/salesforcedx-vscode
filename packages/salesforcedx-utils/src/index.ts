export * from './';
export * from './cli';
export * from './constants';
export * from './helpers';
export * from './i18n';
export * from './requestService';
export * from './types';
