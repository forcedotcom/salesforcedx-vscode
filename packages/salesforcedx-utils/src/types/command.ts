export interface Command {
  readonly command: string;
  readonly description?: string;
  readonly args: string[];
  readonly logName?: string;

  toString(): string;
  toCommand(): string;
}
