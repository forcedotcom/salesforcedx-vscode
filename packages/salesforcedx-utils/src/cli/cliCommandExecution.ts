import { ChildProcess } from 'child_process';
import { Observable } from 'rxjs/Observable';
import { Subscription } from 'rxjs/Subscription';

import { CancellationToken, CommandExecution } from '../types';
import { Command } from './command';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const treeKill = require('tree-kill');
export const NO_PID_ERROR = 'No process associated with sfdx command.';
export const NO_STDOUT_ERROR = 'No stdout found for childProcess';
export const NO_STDERR_ERROR = 'No stderr found for childProcess';
export const CANCELLATION_INTERVAL = 1000;
export const KILL_CODE = 'SIGKILL';

export class CliCommandExecution implements CommandExecution {
  public readonly command: Command;
  public readonly cancellationToken?: CancellationToken;
  public readonly processExitSubject: Observable<number | undefined>;
  public readonly processErrorSubject: Observable<Error | undefined>;
  public readonly stdoutSubject: Observable<Buffer | string>;
  public readonly stderrSubject: Observable<Buffer | string>;

  private readonly childProcessPid: number;

  constructor(
    command: Command,
    childProcess: ChildProcess,
    cancellationToken?: CancellationToken
  ) {
    this.command = command;
    this.cancellationToken = cancellationToken;

    if (!childProcess.pid) {
      throw new Error(NO_PID_ERROR);
    }
    this.childProcessPid = childProcess.pid;

    let timerSubscriber: Subscription | null;

    // Process
    this.processExitSubject = Observable.fromEvent(
      childProcess,
      'exit'
    ) as Observable<number | undefined>;
    this.processExitSubject.subscribe(() => {
      if (timerSubscriber) {
        timerSubscriber.unsubscribe();
      }
    });
    this.processErrorSubject = Observable.fromEvent(
      childProcess,
      'error'
    ) as Observable<Error | undefined>;
    this.processErrorSubject.subscribe(() => {
      if (timerSubscriber) {
        timerSubscriber.unsubscribe();
      }
    });

    // Output
    if (!childProcess.stdout) {
      throw new Error(NO_STDOUT_ERROR);
    }
    this.stdoutSubject = Observable.fromEvent(childProcess.stdout, 'data');
    if (!childProcess.stderr) {
      throw new Error(NO_STDERR_ERROR);
    }
    this.stderrSubject = Observable.fromEvent(childProcess.stderr, 'data');

    // Cancellation watcher
    if (cancellationToken) {
      const timer = Observable.interval(CANCELLATION_INTERVAL);
      timerSubscriber = timer.subscribe(async () => {
        if (cancellationToken.isCancellationRequested) {
          try {
            await this.killExecution();
          } catch (e) {
            console.log(e);
          }
        }
      });
    }
  }

  public async killExecution(signal = KILL_CODE) {
    return killPromise(this.childProcessPid, signal);
  }
}

/**
 * This is required because of https://github.com/nodejs/node/issues/6052
 * Basically if a child process spawns it own children  processes, those
 * children (grandchildren) processes are not necessarily killed
 */
async function killPromise(processId: number, signal: string): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    treeKill(processId, signal, (err: Error | undefined) => {
      err ? reject(err) : resolve();
    });
  });
}
