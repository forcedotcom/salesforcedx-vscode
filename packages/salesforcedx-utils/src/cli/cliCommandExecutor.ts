/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { SpawnOptions } from 'child_process';
import 'rxjs/add/observable/fromEvent';
import 'rxjs/add/observable/interval';

import { TELEMETRY_HEADER } from '../constants';
import { CancellationToken, Command } from '../types';
import { CliCommandExecution } from './cliCommandExecution';
import { GlobalCliEnvironment } from './globalCliEnvironment';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const cross_spawn = require('cross-spawn');

export class CliCommandExecutor {
  protected static patchEnv(
    options: SpawnOptions,
    baseEnvironment: Map<string, string>
  ): SpawnOptions {
    // start with current process environment
    const env = Object.create(null);

    // inherit current process environment
    Object.assign(env, process.env);

    // now push anything from global environment
    baseEnvironment.forEach((value, key) => {
      env[key] = value;
    });

    if (env) {
      env.SFDX_TOOL = TELEMETRY_HEADER;
    }

    // then specific environment from Spawn Options
    if (typeof options.env !== 'undefined') {
      Object.assign(env, options.env);
    }

    options.env = env;
    return options;
  }

  private readonly command: Command;
  private readonly options: SpawnOptions;

  public constructor(
    command: Command,
    options: SpawnOptions,
    inheritGlobalEnvironmentVariables = true
  ) {
    this.command = command;
    this.options = inheritGlobalEnvironmentVariables
      ? CliCommandExecutor.patchEnv(
          options,
          GlobalCliEnvironment.environmentVariables
        )
      : options;
  }

  public execute(cancellationToken?: CancellationToken): CliCommandExecution {
    const childProcess = cross_spawn(
      this.command.command,
      this.command.args,
      this.options
    );
    return new CliCommandExecution(
      this.command,
      childProcess,
      cancellationToken
    );
  }
}
