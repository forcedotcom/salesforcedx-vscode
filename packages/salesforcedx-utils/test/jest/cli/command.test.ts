import { CommandBuilder } from '../../../src';
import { Command } from '../../../src/cli/command';

describe('Command unit tests.', () => {
  const testCommand = 'iAmACommand';

  it('Should be able to create a command.', () => {
    const commandBuilderInst = new CommandBuilder(testCommand);
    const command = new Command(commandBuilderInst);
    expect(command).toBeInstanceOf(Command);
    expect(command.command).toEqual(testCommand);
    expect(command.description).toBeUndefined();
    expect(command.args).toEqual([]);
    expect(command.logName).toBeUndefined();
  });

  it('Should be able to pass a fully built builder.', () => {
    const commandBuilderInst = new CommandBuilder(testCommand)
      .withDescription('test description')
      .withArg('immaArg')
      .withFlag('--silent', 'quiet')
      .withJson()
      .withLogName('testLogName');

    const command = new Command(commandBuilderInst);
    expect(command).toBeInstanceOf(Command);
    expect(command.command).toEqual(commandBuilderInst.command);
    expect(command.description).toEqual(commandBuilderInst.description);
    expect(command.args).toEqual(commandBuilderInst.args);
    expect(command.logName).toEqual(commandBuilderInst.logName);
  });

  it('Should return description for toString if exists.', () => {
    const testDescription = 'super cool descrption';
    const commandBuilderInst = new CommandBuilder(testCommand);
    commandBuilderInst.withDescription(testDescription);
    const command = new Command(commandBuilderInst);
    expect(command.toString()).toEqual(testDescription);
  });

  it('Should construct string if no description provided.', () => {
    const commandBuilderInst = new CommandBuilder(testCommand)
      .withArg('immaArg')
      .withFlag('--silent', 'quiet')
      .withJson()
      .withLogName('testLogName');

    const command = new Command(commandBuilderInst);
    expect(command.toString()).toMatchSnapshot();
  });

  it('Should build a command.', () => {
    const commandBuilderInst = new CommandBuilder(testCommand)
      .withDescription('test description')
      .withArg('immaArg')
      .withFlag('--silent', 'quiet')
      .withJson()
      .withLogName('testLogName');

    const command = new Command(commandBuilderInst);
    expect(command.toCommand()).toMatchSnapshot();
  });
});
