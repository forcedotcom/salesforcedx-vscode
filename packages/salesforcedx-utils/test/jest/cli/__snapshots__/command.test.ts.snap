// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command unit tests. Should build a command. 1`] = `"iAmACommand immaArg --silent quiet --json --loglevel fatal"`;

exports[`Command unit tests. Should construct string if no description provided. 1`] = `"iAmACommand immaArg --silent quiet --json --loglevel fatal"`;
